Add Fetch API demos

addyosmani opened the pull request:
reviewers: @gauntface @jeffposnick

addyosmani committed:
Add Fetch API demos

gauntface commented:
LGTM. Think samples are solid - are you still struggling with the POST request or is your one working?

Realised my sample is bust thanks for a change in Chrome such that no-cors mode has been disabled due to a security concern.

jeffposnick commented on fetch-api/fetch-post.html:
```diff
@@ -0,0 +1,119 @@
+        });
+    }
+
+    var jsonValidator = 'http://validate.jsontest.com/';
```
You'll get a mixed content failure when the sample page is accessed via HTTPS.

`Mixed Content: The page at 'https://rawgit.com/GoogleChrome/samples/fetch/fetch-api/fetch-post.html' was loaded over HTTPS, but requested an insecure resource 'http://validate.jsontest.com/'. This request has been blocked; the content must be served over HTTPS.`

And it looks like https://validate.jsontest.com/ isn't a thing.

addyosmani replied:
I understand the mixed content warning, but it looks like the validation service I was using is now down. I'll see if I can find a suitable replacement that we can post data to without mixed content warnings.

jeffposnick commented on fetch-api/fetch-reddit.html:
```diff
@@ -0,0 +1,118 @@
+            .then(function(response) {
+                return response.json()
+            }).then(function(json) {
+                console.log(json);
```
Did you intend to keep this, or was it for debugging during development?

addyosmani replied:
I intended to keep this to demonstrate the complete payload, but didn't want to log to the page. If it's better to drop these to be consistent with other demos happy to do so.

jeffposnick commented on fetch-api/fetch-response-metadata.html:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    fetch('users.json').then(function(response) {
+        console.log(response);
```
Did you intend to keep this, or was it for debugging during development?

jeffposnick commented on fetch-api/fetch-reddit.html:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    fetch('https://www.reddit.com/reddits.json')
```
This returned some subreddits of questionable taste when I ran the sample.

I'd recommend an API call that will give a more curated response, like http://www.reddit.com/api/subreddits_by_topic.json?query=javascript

addyosmani replied:
That makes sense. I'll update this to provide a slightly more predictable quality of results. Thanks Jeff!

jeffposnick commented on fetch-api/fetch-response-metadata.html:
```diff
@@ -0,0 +1,88 @@
+        log(response.status);
+        log(response.statusText);
+        log(response.type);
+        log(response.url);
```
Move this to the top of the log.

addyosmani replied:
SGTM

jeffposnick commented on fetch-api/fetch-response-metadata.html:
```diff
@@ -0,0 +1,88 @@
+
+    fetch('users.json').then(function(response) {
+        console.log(response);
+        log(response.headers.get('Content-Type'));
```
Could you add the string `'Content-Type: '` prior to logging the value? Ditto for the other lines. It would make it clearer what was being logged without needing to match up lines in the `view-source:`.

addyosmani replied:
Good call. That will make it more clear what's being output. It'll land with the next set of commits.

jeffposnick commented on fetch-api/fetch-success-error-handlers.html:
```diff
@@ -0,0 +1,98 @@
+        return response.json()
+    }
+
+    fetch('users.json')
```
How about adding in another, failing `fetch()` call with the same logic? Request `'invalid.json'` or something.

addyosmani replied:
Good idea!

gauntface commented:
Mega nit pick - any chance of a index.html to make it easier to navigate these demos or will gh-pages do this?

addyosmani commented:
@gauntface I thought about that, but added in a README index that we'll be pointing to which lists up to all the samples instead. See: https://github.com/GoogleChrome/samples/blob/823dd3edc6816c6a1bd5b1bfba1a69568ec64f28/fetch-api/README.md. We can easily turn that into an index if needed though.

gauntface commented:
Looks good.

addyosmani committed:
Address Fetch API feedback

addyosmani commented:
@jeffposnick Addressed your feedback. I've also refactored the POST example to use the GitHub Gist API to create a new gist based on user input. It was one of the few open-APIs that supported unauthenticated POST requests :)

jeffposnick commented:
:+1:
